toy_2d/src/ilqr.py

```diff
@@ -0,0 +1,333 @@
+        :param dd: list of "feed-forward" components of iLQR update, should be length N-1
+        :param KK: list of "Feedback" LQR gain components of iLQR update, should be length N-1
+        :return: A tuple (xtraj, utraj) containing the updated state and input
+                 trajectories after applying the iLQR forward pass
+        """
+
+        xtraj = [np.zeros((self.nx,))] * self.N
+        utraj = [np.zeros((self.nu,))] * (self.N - 1)
+        xtraj[0] = xx[0]
+
+        # TODO: compute forward pass
+        for i in range(self.N-1):
+            utraj[i] = uu[i] + KK[i]@(xtraj[i] - xx[i]) + lr*dd[i]
+            # if(utraj[i][0]<0 and utraj[i][1]>0): utraj[i] = np.array([0,0])
+            #TO DO
+            # xtraj[i+1] = quad_sim.F(xtraj[i], utraj[i], self.dt)
+            xtraj[i+1] = self.sim_forward(xtraj[i], utraj[i])
+        return xtraj, utraj
+
+    def get_linearized_discrete_dynamics(self, state, action):
+        q = np.array([state[0], state[2], state[4]])
+        v = np.array([state[1], state[3], state[5]])
+        self.lcs.set_linearization_point(q,v, action)
+        A, B_gen = self.lcs.get_A_matrix(), self.lcs.get_B_matrix()
+        pmat = self.lcs.get_P_matrix()
+        B = B_gen@pmat
+        C_mat = self.lcs.get_C_matrix()
+        G_mat = self.lcs.get_G_matrix()
+        J_mat = self.lcs.get_J_matrix()
+        H_mat = self.lcs.get_H_matrix()
+        #access original dynamics to get the original lambdas
+        curr_lam = self.get_lamda(state, action)
+        eps = 1e-3
+        # check if there is a contact at all. if not use the normal dynamics itself as cube is in air
+        if(np.argwhere(curr_lam>eps).shape[0]>0):
+            curr_lam = curr_lam[curr_lam>eps]
+            # print(curr_lam)
+            C_mat = C_mat[:, np.where(curr_lam>eps)[0]]
+            J_mat = J_mat[:, np.where(curr_lam>eps)[0]]
+            J_pinv = np.linalg.pinv(J_mat)
+            A_d, B_d = -C_mat@J_pinv@G_mat, -C_mat@J_pinv@H_mat@pmat
+            A, B = A + A_d, B+B_d
+
+        return A,B
+    
+    def get_lamda(self, state, action):
+        self.lcs.system.set_initial_state(state)
+        self.lcs.system.step_dynamics(action)
+        return self.lcs.system.lambda_history[-1,:]
+
+    def backward_pass(self,  xx, uu):
+        """
+        :param xx: state trajectory guess, should be length N
+        :param uu: input trajectory guess, should be length N-1
+        :return: KK and dd, the feedback and feedforward components of the iLQR update
+        """
+        dd = [np.zeros((self.nu,))] * (self.N - 1)
+        KK = [np.zeros((self.nu, self.nx))] * (self.N - 1)
+        # TODO: compute backward pass
+        V_xp1 = np.zeros((self.nx)) #jacobian of value func at the next state
+        V_xxp1 = np.zeros((self.nx, self.nx)) #hess of value func at the next state
+        for i in range(self.N-1,-1,-1):
+            # print(i)
+            state = xx[i]
+            if(i == self.N-1):
```

Comment:
You can eliminate this `if` statement if you instead make these assignments for `V_xp1` and `V_xxp1` in lines 208 and 209, instead of there setting them to all zeros.